crates/volta-core/src/tool/pnpm/fetch.rs

```diff
@@ -0,0 +1,184 @@
+
+    let progress = progress_bar(
+        archive.origin(),
+        &tool_version("pnpm", version),
+        archive
+            .uncompressed_size()
+            .unwrap_or_else(|| archive.compressed_size()),
+    );
+    let version_string = version.to_string();
+
+    archive
+        .unpack(temp.path(), &mut |_, read| {
+            progress.inc(read as u64);
+        })
+        .with_context(|| ErrorKind::UnpackArchiveError {
+            tool: "pnpm".into(),
+            version: version_string.clone(),
+        })?;
+
+    let bin_path = temp.path().join("package").join("bin");
+    overwrite_launcher(&bin_path, "pnpm")?;
+    overwrite_launcher(&bin_path, "pnpx")?;
+
+    #[cfg(windows)]
+    {
+        overwrite_cmd_launcher(&bin_path, "pnpm")?;
+        overwrite_cmd_launcher(&bin_path, "pnpx")?;
+    }
+
+    let dest = volta_home()?.pnpm_image_dir(&version_string);
+    ensure_containing_dir_exists(&dest)
+        .with_context(|| ErrorKind::ContainingDirError { path: dest.clone() })?;
+
+    rename(temp.path().join("package"), &dest).with_context(|| ErrorKind::SetupToolImageError {
+        tool: "pnpm".into(),
+        version: version_string.clone(),
+        dir: dest.clone(),
+    })?;
+
+    progress.finish_and_clear();
+
+    // Note: We write this after the progress bar is finished to avoid display bugs with re-renders of the progress
+    debug!("Installing pnpm in '{}'", dest.display());
+
+    Ok(())
+}
+
+/// Return the archive if it is valid. It may have been corrupted or interrupted in the middle of
+/// downloading.
+// ISSUE(#134) - verify checksum
+fn load_cached_distro(file: &Path) -> Option<Box<dyn Archive>> {
+    if file.is_file() {
+        let file = File::open(file).ok()?;
+        Tarball::load(file).ok()
+    } else {
+        None
+    }
+}
+
+/// Determine the remote URL to download from, using the hooks if avaialble
+fn determine_remote_url(version: &Version, hooks: Option<&ToolHooks<Pnpm>>) -> Fallible<String> {
+    let version_str = version.to_string();
+    match hooks {
+        Some(&ToolHooks {
+            distro: Some(ref hook),
+            ..
+        }) => {
+            debug!("Using pnpm.distro hook to determine download URL");
+            let distro_file_name = Pnpm::archive_filename(&version_str);
+            hook.resolve(version, &distro_file_name)
+        }
+        _ => Ok(public_registry_package("pnpm", &version_str)),
+    }
+}
+
+/// Fetch the distro archive from the internet
+fn fetch_remote_distro(
+    version: &Version,
+    url: &str,
+    staging_path: &Path,
+) -> Fallible<Box<dyn Archive>> {
+    debug!("Downloading {} from {}", tool_version("pnpm", version), url);
+    Tarball::fetch(url, staging_path).with_context(download_tool_error(
+        tool::Spec::Pnpm(VersionSpec::Exact(version.clone())),
+        url,
+    ))
+}
+
+/// Create executable launchers for the pnpm and pnpx binaries
+fn overwrite_launcher(base_path: &Path, tool: &str) -> Fallible<()> {
```

Comment:
small naming nit: It appears that `pnpm` doesn't include launchers at all by default in the package tarball, so rather than "overwrite_[cmd_]launcher", would it make sense for these to be named "write_[cmd_]launcher"?